doc/arch/adr-002-digest-mvp.md

```diff
@@ -0,0 +1,187 @@
+
+##### Shall
+
+- Be interfaced by a scheduling service to run at a set time appropriate
+  for their period
+- Create a [`DigestRun`](#digestrun) model
+- Ensure that multiple digests for the same period are not run concurrently
+
+##### Should
+
+- Be configurable as to what times a digest runs from and until
+
+#### `DigestRun`
+
+A model which represents a distinct run of a period of digest (daily or
+weekly).
+
+Has a responsibility to persist data that is related to a full digest run for
+all subscribers.
+
+##### Shall
+
+- Store information such as
+  - Date digest is run on
+  - The period the digest is for
+  - The start time for the digest period
+  - The end time for the digest period
+
+##### Should
+
+- Store a timestamp for when all the emails were created for the digest, to act
+  as indicator it is complete.
+
+##### May
+
+- Have a unique index for digest date and period, however this would limit
+  ability to re-run a digest in the case of a problem.
+
+#### `DigestSubscriberQuery`
+
+Responsible for taking a [`DigestRun`](#digestrun) instance and using that to
+determine which subscribers should receive an email for the digest period
+
+##### Shall
+
+- Use a `DigestRun` to determine which subscribers should receive an email for
+  a digest period
+
+##### Should
+
+- Return a list of ids of subscribers
+
+##### May
+
+- Return an ActiveRecord scope to allow the caller to deal with this returning
+  an unlimited output.
+
+#### `DigestRunSubscriber`
+
+A model which is used to associate a [`DigestRun`](#digestrun) with a subscriber
+instance. Will mostly be used as a persisted piece of data that logs which
+subscribers receive an email due to a digest. It is unlikely to be needed
+long term.
+
+##### Shall
+
+- Store
+  - An association to `DigestRun`
+  - An association to `Subscriber`
+
+##### Should
+
+- Store a timestamp for when it has been completed, can be used therefore to
+  work out if a `DigestRun` is complete or not.
+
+#### `DigestGenerationWorker`
+
+A worker entity that accepts the argument of a
+[`DigestRunSubscriber`](#digestrunsubscriber) and has the responsibility to
+create an Email entity for the associated subscriber.
+
+##### Shall
+
+- Interface with [`ContentChangeQuery`](#contentchangequery) to load content
+  changes
+- Interface with [`DigestEmailRenderer`](#digestemailrenderer) to render email
+  content
+- Create an `Email` entity
+- Create a `SubscriptionContent` entity, which will require knowing which
+  subscription an email is for
+
+##### Should
+
+- Mark a `DigestRunSubscriber` as completed
+- Mark `DigestRun` as completed if all `DigestRunSubscriber` are complete for
+  the `DigestRun`
+
+##### May
+
+- Liaise with a `DigestGeneratorService` to do the work if the code for the
+  worker becomes complicated or should be more re-usable
+
+#### `ContentChangeQuery`
+
+For a given [`DigestRunSubscriber`](#digestrunsubscriber) this determines which,
```

Comment:
I thought it might be good to pass in the subscriptions to this class, since they're easy to get from the `DigestRunSubscriber`, and it means the `ContentChangeMatcher` and `SubscriptionMatcher` classes are exact reverses of each other.

Reply:
I just had a chat with @gpeng about this sort of stuff and hadn't properly considered your comment.

We were wrestling with how we'd know which subscriptions are associated with a content change. We were toying with:

`ContentChangeMatcher.call(digest_run_subscriber)` which returns something like
```
{ 
  1 => [content_change, content_change],
  2 => [content_change],
}
```
Where 1 and 2 are subscription ids

In theory we could do this differently with more steps in the worker class like
```
subscriptions = digest_run_subscriber.subscriber.subscriptions
grouped_content_changes = subscriptions.map do |s| 
  {
     subscription: s,
     content_changes: ContentChangeMatcher.call(s.subscriber_list)
  }
end
```

But I don't know for sure what structure we want to go into the `DigestEmailRenderer` so will probably leave this with options in this doc.